src/scheduler/Executor.cpp

```diff
@@ -79,99 +119,116 @@ void Executor::executeTasks(std::vector<int> msgIdxs,
     // Iterate through and invoke tasks
     for (int msgIdx : msgIdxs) {
         const faabric::Message& msg = req->messages().at(msgIdx);
-        int threadPoolIdx = msg.appindex() % threadPoolSize;
+
+        // If executing threads, we must always keep thread pool index zero
+        // free, as this may be executing the function that spawned them
+        int threadPoolIdx;
+        if (isThreads) {
+            assert(threadPoolSize > 1);
+            threadPoolIdx = (msg.appindex() % (threadPoolSize - 1)) + 1;
```

Comment:
As discussed offline, each executor will only ever be handling one function, and the threads in this thread pool are only for threads spawned by that function. We rely on the threading implementation to set the `appindex` properly, but there are definitely some potential bugs in here once things get overloaded (although the scheduler will do its best to avoid this).